### DISCUSSION TITLE
Q&A: ReptileMetaLearner configuration — inner_steps, meta_lr, EWC integration, and few-shot convergence (Phase 7.1)

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#261] [Q&A]
## Questions and answers about `ReptileMetaLearner` configuration, hyperparameters, and few-shot adaptation

This Q&A covers common questions about setting up Phase 7.1 Reptile meta-learning in ASI:BUILD.

---

**Q1: How do I choose `inner_steps` (k)?**

`inner_steps` controls how far the inner loop travels from the meta-parameters before computing the update direction. Trade-offs:

| `k` | Effect | When to use |
|---|---|---|
| 1 | Minimal adaptation; almost no meta-learning signal | Debugging / sanity check only |
| 5 | Standard Reptile (original paper default) | Most production use-cases |
| 10 | Stronger adaptation; slower meta-update | Long-horizon tasks with many transitions |
| 20+ | Risk of over-fitting support set; noisy meta-gradients | Avoid unless task is very stationary |

**Default**: `inner_steps=5` matches the original Reptile paper and works well with the SLEEP_PHASE budget.

---

**Q2: What is `meta_lr` (ε) and how should I tune it?**

`meta_lr` is the interpolation coefficient between current meta-params and task-adapted params. It is **not** a standard learning rate — it is a weight on how much each meta-update shifts the initialisation point.

| `meta_lr` | Behaviour |
|---|---|
| 0.001 | Very slow meta-adaptation; safe for online deployment |
| 0.01 | Moderate; recommended starting point |
| 0.1 | Fast adaptation; can cause oscillation if task distribution is non-stationary |
| 1.0 | Replaces meta-params with task-adapted params entirely (no interpolation) |

Start with `meta_lr=0.01` and increase if few-shot performance plateaus after 100+ SLEEP cycles.

---

**Q3: Can `ReptileMetaLearner` be used without EWC?**

Yes. `ewc_regulariser` is optional:

```python
# Without EWC (Phase 6 not required):
meta_learner = ReptileMetaLearner(base_learner=my_learner, meta_lr=0.01)

# With EWC constraint in inner loop:
meta_learner = ReptileMetaLearner(
    base_learner=my_learner,
    meta_lr=0.01,
    ewc_regulariser=ewc,           # from Phase 6.3
    fisher_accumulator=accumulator, # from Phase 6.4
)
```

When EWC is absent, the inner loop is unconstrained SGD — appropriate if starting fresh without a pre-trained backbone.

---

**Q4: How does `task_batch_size` affect meta-learning quality?**

`task_batch_size` determines how many tasks are sampled per SLEEP cycle and averaged into a single meta-update.

| Batch size | Variance | Compute cost | Notes |
|---|---|---|---|
| 1 | High | Minimal | Online Reptile; noisy but fast |
| 4 | Moderate | Low | Good default |
| 8 | Low | Medium | Preferred if TaskRegistry has 8+ active tasks |
| 16+ | Very low | High | Only viable for large agent fleets |

If `TaskRegistry` has fewer tasks than `task_batch_size`, `ReptileMetaLearner` will sample with replacement.

---

**Q5: How does `snapshot_meta_params` / `restore_meta_params` work?**

`snapshot_meta_params()` returns a deep copy of the current meta-parameters as a `MetaParamSnapshot`. Use it before risky meta-updates to allow rollback:

```python
snap = meta_learner.snapshot_meta_params()
result = await meta_learner.meta_update(risky_task_batch)
if result.meta_loss > threshold:
    meta_learner.restore_meta_params(snap)  # rollback
```

Note: `restore_meta_params` does **not** reset `self._step` — the step counter is monotonic for Prometheus consistency.

---

**Q6: Does Reptile work with the existing `TaskRegistry` from Phase 6.5?**

Yes — `ReptileMetaLearner` uses `TaskContextManager` from Phase 6.5 to set the `task_id` context variable during each inner loop. This ensures `FisherAccumulator` and any `STDPOnlineLearner` calls route to the correct per-task Fisher matrix automatically.

No changes to `TaskRegistry` are needed for Phase 7.1.

---

**Q7: How many SLEEP cycles are needed before Reptile improves few-shot accuracy?**

Empirically, Reptile typically shows measurable improvement after **50–100 meta-update steps** (50–100 SLEEP cycles). You can monitor convergence via:

```promql
# Meta-loss trend (should decrease)
asi_reptile_meta_loss

# Update norm (should decrease as meta-params converge)
asi_reptile_update_norm

# Inner-loop loss distribution (should tighten)
histogram_quantile(0.95, rate(asi_reptile_inner_loss_bucket[1h]))
```

Related: Issue #259 · Discussion #260 (Show & Tell) · Discussion #258 (Phase 7 planning)
